Move the Ansible Tower UI to the Automation tab

lgalis opened the pull request:
Add the Ansible Tower UI to the Automation Tab.

-----------------------
Depends on https://github.com/ManageIQ/manageiq/pull/13526 - will remove the WIP label after 13526 is merged.

This PR moves the Ansible Tower UI from the Configuration tab to the Automation Tab.
It uses the new AutomationManager model for Ansible introduced in PR #13630. 

Links:
-------------------------------------------

https://www.pivotaltracker.com/n/projects/1937457

![screenshot from 2017-01-25 01-17-12](https://cloud.githubusercontent.com/assets/12769982/22280002/107844ee-e29c-11e6-9e0f-5e843816f191.png)

![screenshot from 2017-01-30 01-31-19](https://cloud.githubusercontent.com/assets/12769982/22414234/e5d6a746-e68b-11e6-9bf4-9024fab4a699.png)

![screenshot from 2017-01-30 22-39-44](https://cloud.githubusercontent.com/assets/12769982/22453564/b6159bbc-e74d-11e6-8ada-ada7074bf163.png)
![screenshot from 2017-01-30 22-38-52](https://cloud.githubusercontent.com/assets/12769982/22453565/b6158ef6-e74d-11e6-9679-244760efee8f.png)

![screenshot from 2017-02-09 13-08-43](https://cloud.githubusercontent.com/assets/12769982/22796614/056a537e-eec9-11e6-9296-f66ea9eaf4b7.png)

lgalis commented:
@miq-bot add_label wip

chessbyte commented on app/controllers/application_controller/ci_processing.rb:
```diff
@@ -1929,7 +1929,7 @@ def process_objects(objs, task, display_name = nil)
   end
 
-  def foreman_button_operation(method, display_name)
+  def configmgr_button_operation(method, display_name)
```
Please rename `configmgr` to `configuration_manager`

lgalis commented:
@dclarizio, @h-kataria - please review

lgalis commented:
@dclarizio , @h-kataria - please review  - Please note that this PR needs  ManageIQ/manageiq#13526.

dclarizio commented:
@martinpovolny added you as a reviewer to check for any of the currenty UI refactoring patterns that @lgalis could apply either in this PR or in a follow up.

lgalis commented:
@martinpovolny - just FYI - the AnsibleTower::ConfigurationManager  will be changed to AnsibleTower::AutomationManager after the PR 13630 is merged.

lgalis commented:
@miq-bot remove_label wip

lgalis commented:
@martinpovolny - Please review - this PR was updated to use the new AutomationManager model. 
The spec failures are due to the fact that the PR with the new features, https://github.com/ManageIQ/manageiq/pull/13526 is not yet merged.

martinpovolny commented:
@lgalis : To begin there's a lot of code climate issues in the javascript.

Such as:
> You should not set properties on $scope in controllers. Use controllerAs syntax and add data to "this"

Conversion to the `controllerAs` syntax is pretty straightforward, please, use it. If you don't, we introduce new technical debt as we add new code.

CC plugin does now work with this PR, because it's so huge.

martinpovolny commented:
`automation_manager_button_operation` is a 1:1 c-n-p from app/controllers/application_controller/ci_processing.rb except for the function called at the very end. 

Please, extract a method and parametrize it instead of c-n-p.

`process_automation_manager` also looks very similar to it's `CiProcessing` counterpart.

Please, allign the toobar files the same way as e.g. `app/helpers/application_helper/toolbar/miq_ae_class_center.rb`. I know that rubocop will complain, but it looks much better w/o the biiiiiiiiig indentation, it fits the screen and is readable.....

martinpovolny commented:
I need to go much deeper with this one but at least you are not blocked on the review here.

martinpovolny commented on app/controllers/automation_manager_controller.rb:
```diff
@@ -0,0 +1,1025 @@
+
+  def tagging
+    case x_active_accord
+    when :automation_manager_providers
```
you are testing 2 different priviledges, but then calling the same method with the same arguments

is that intentional?

lgalis replied:
The configured systems are displayed on 2 different accordions - each with it's own privilege ( this is the same as in the foreman_provider)

martinpovolny replied:
I uderstand that it's displayed in 2 places. What surprises me is that there are 2 different RBAC features tested for the same method call. This looks wrong or at least strange.

martinpovolny commented on app/controllers/automation_manager_controller.rb:
```diff
@@ -0,0 +1,1025 @@
+    @explorer = true if request.xml_http_request? # Ajax request means in explorer
+
+    @gtl_url = "/show"
+    set_summary_pdf_data if "download_pdf" == @display
```
then new controller supports "download_pdf"?

martinpovolny commented:
I wonder if we can split this somehow to get it in and to allow a better review.

The textual summaries look ok, the toolbars would look ok if you fix the allignment.

Need to test if the menus + the feaure additions in the core repo PR don't mess up the RBAC editor.

We can surely merge the .svg image. 

The javascripts have a load of style issues (i poited out the controllerFor above). Can you check those on CodeClimate? Most should be pretty quick to fix.

The controller code needs a detailed review.

lgalis commented:
@martinpovolny  - I addressed everything you mentioned, except the javascript - the controllerAs use would make this controller different than all the other angular controllers we have at the moment. Do we want this refactoring here or as a separate PR? 
I have added screenshots for the RBAC tree with the new features.

martinpovolny commented:
@lgalis : what was the solution for the pdf stuff? I can see the string "pdf" in the changes.

lgalis commented:
@martinpovolny  - the pdf - I removed the line to set the summary data when display=pdf as that was no longer necessary once Zita's PR for download_pdf was merged

martinpovolny commented:
@lgalis : I don't get it, is the download_pdf action supported by the new controller or not? Because I can see code that deals with it.

On the `controllerAs` suff: I understand that we have other controllers that still use `$scope`. But I don't understand the reason to add more of such code. It represents a technical debt. Why are we introducing new technical debt when adding new code? Is it absolutely necessary?

I can see that you want that in now as it is, because there's some pressure to deliver. Sure we can fit that in a follow up PR. But that's the case all the time with all the PRs by all the teams.

martinpovolny commented on app/views/automation_manager/explorer.html.haml:
```diff
@@ -0,0 +1,20 @@
+
+-# These are the initial divs that will go inside center_cell_div
+- if @inventory_group_record
+  #main_div
```
seems all the branches add the `main_div`, can we move it one up?

martinpovolny commented:
@lgalis: so I finally got a chance to take a look at the controller.

I actually did `vim -d app/controllers/provider_foreman_controller.rb app/controllers/automation_manager_controller.rb`

Seems that `provider_foreman_controller` is where most of the code comes from. There where some changes to the foreman controller why you worked on your copy that where not reflected, I can see the 'GenericSessionMixins' is not used in your new controller.

Generally I'd say you should move all the reused methods to a mixin and include that mixin in both controllers. In places where you do minimal changes (a lot of places) all you need to do is parametrize the methods. 

An example: `@sb[:foreman_search_text] vs @sb[:automation_manager_search_text]` there's already a this:
```
     def get_session_data
     prefix      = self.class.session_key_prefix 
     ...
``` 
in the `GenericSessionMixin` so no need to c-n-p `accordion_select`, just make use of the prefix.

Another example: `show` is the same except for the recoed assignment. So all you need is have a method such as `identify_record(id)` and put the special code there.

Then `@provider_cfgmgmt` can be renamed to `@provider` I guess and more code can be shared.

I can carry on, all you need is run a diff side by side in your favorite editor and move a method by method to a mixin.

Does it make sense?

lgalis commented:
@martinpovolny - This is only the first step in separating the Ansible tower from the Foreman provider controller. I did have it on my list to create a mixin that can be used in both controllers - I wanted to have that in a follow up PR as time allows ( the Pivotal story states that as well).

lgalis commented:
@martinpovolny -  I'll mark the PR as WIP and make the changes for the angular controller as well as the mixin

martinpovolny commented:
@lgalis : Thanks. I'd really prefer to have the changes done before merging this. Ping me, I'll try to help if you have any problem.

lgalis commented:
@martinpovolny - I addressed most of your comments - except the use of the controllerAs in the angular code. In order to use this, changes are required to other controllers that we use in multiple places, like teh credentials controller, as well as to common code we use for buttons and checking change. I think this would be more than the scope of this PR. I think we need to first make the changes needed to the shared controllers and/or directives/services as needed then we can use the controllerAs in the new controllers/

lgalis commented:
@miq-bot remove_label wip

himdel commented:
Ping @epwinchell .. can you check the icon renames please?

epwinchell commented:
@lgalis @himdel Could we just rename the icon "vendor-ansible.svg"? 

We have a tendency to get too specific with the names and then later make multiple copies.

himdel commented:
@epwinchell .. assuming we can make the code return that.. note that the vendor images are determined quite ..randomly..

```
svg/vendor-#{@object.image_name}.svg
svg/vendor-#{@object.top_level_namespace.downcase}.svg
svg/vendor-#{@record.configuration_manager.image_name}.svg
svg/vendor-#{@root.image_name}.svg
svg/vendor-#{@vat_tree.instance_variable_get(:@root).image_name}.svg
svg/vendor-#{@vmminfo[0][:description].downcase}.svg
svg/vendor-#{ems.image_name}.svg
svg/vendor-#{h(@drift_obj.vendor.downcase)}.svg
svg/vendor-#{h(@drift_obj.vmm_vendor_display.downcase)}.svg
svg/vendor-#{h(@host.vmm_vendor_display.downcase)}.svg
svg/vendor-#{h(item.image_name)}.svg
svg/vendor-#{h(item.vendor)}.svg
svg/vendor-#{h(item.vendor.downcase)}.svg
svg/vendor-#{h(item.vmm_vendor_display.downcase)}.svg
svg/vendor-#{h[:vendor].downcase}.svg
svg/vendor-#{h[:vmm_vendor].downcase}.svg
svg/vendor-#{image_name.downcase}.svg
svg/vendor-#{image_name}.svg
svg/vendor-#{lives_on_ems.image_name}.svg
svg/vendor-#{o.image_name}.svg
svg/vendor-#{object.image_name}.svg
svg/vendor-#{provider_type.to_s.underscore.downcase}.svg
svg/vendor-#{record.top_level_namespace.downcase}.svg
svg/vendor-#{vendor.downcase}.svg
svg/vendor-#{vendor}.svg
svg/vendor-#{vmm_vendor_display.downcase}.svg
```

lgalis commented:
@epwinchell - the image was renamed, not duplicated - to reflect the new model/location for the Ansible Tower

epwinchell commented:
@lgalis I am aware of that - it was a general statement

himdel commented:
When adding euwe/yes or euwe/no, please add the same to https://github.com/ManageIQ/manageiq-ui-classic/pull/350

martinpovolny commented:
The textual summary helper functions seem to added here seem to contain a lot of dead code :-(
